### PR TITLE
1.3.4.1 Update (Subfolders)

### PR DESCRIPTION
**Pull request by Zaicon**

### CODE REVIEW

File: Terraria/IO/WorldFile.cs [hunk 1]
```diff
@@ -321,7 +358,7 @@ public static bool GetWorldDifficulty(string WorldFileName)
 						{
 							binaryReader.BaseStream.Position += 20L;
 						}
-						if (num >= 112 && num <= Main.curRelease)
+						if (num >= 112 && num <= 184)
```
Comment (QuiCM) [inline Terraria/IO/WorldFile.cs]: Should be Main.curRelease

File: Terraria/IO/WorldFile.cs [hunk 2]
```diff
@@ -362,7 +399,7 @@ public static string GetWorldName(string WorldFileName)
 					using (BinaryReader binaryReader = new BinaryReader(fileStream))
 					{
 						int num = binaryReader.ReadInt32();
-						if (num > 0 && num <= Main.curRelease)
+						if (num > 0 && num <= 184)
```
Comment (QuiCM) [inline Terraria/IO/WorldFile.cs]: Also should be Main.curRelease

File: Terraria/IO/WorldFile.cs [hunk 3]
```diff
@@ -1103,7 +1161,7 @@ public static void loadWorld()
 		public static int LoadWorld_Version1(BinaryReader fileIO)
 		{
 			int num = WorldFile.versionNumber;
-			if (num > Main.curRelease)
+			if (num > 184)
```
Comment (QuiCM) [inline Terraria/IO/WorldFile.cs]: Main.curRelease

File: Terraria/IO/WorldFile.cs [hunk 4]
```diff
@@ -2054,7 +2112,7 @@ private static int SaveFooter(BinaryWriter writer)
 		private static int SaveHeaderPointers(BinaryWriter writer, int[] pointers)
 		{
 			writer.BaseStream.Position = (long)0;
-			writer.Write(Main.curRelease);
+			writer.Write(184);
```
Comment (QuiCM) [inline Terraria/IO/WorldFile.cs]: Main.curRelease

File: Terraria/IO/WorldFile.cs [hunk 5]
```diff
@@ -2650,16 +2713,12 @@ public static bool validateWorld(BinaryReader fileIO)
...
-			{
-				WorldGen.genRand = new Random((int)DateTime.Now.Ticks);
-			}
 			bool result;
 			try
 			{
 				Stream baseStream = fileIO.BaseStream;
 				int num = fileIO.ReadInt32();
-				if (num == 0 || num > Main.curRelease)
+				if (num == 0 || num > 184)
```
Comment (QuiCM) [inline Terraria/IO/WorldFile.cs]: Main.curRelease

File: Terraria/Utilities/NPCUtils.cs [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+
+		public static NPCUtils.TargetSearchResults SearchForTarget(NPC searcher, Vector2 position, NPCUtils.TargetSearchFlag flags = NPCUtils.TargetSearchFlag.All, NPCUtils.SearchFilter<Player> playerFilter = null, NPCUtils.SearchFilter<NPC> npcFilter = null)
+		{
+
+			float num = 3.40282347E+38f;
+			int nearestNPCIndex = -1;
+			float num2 = 3.40282347E+38f;
+			float nearestTankDistance = 3.40282347E+38f;
+			int nearestTankIndex = -1;
+			NPCUtils.TargetType tankType = NPCUtils.TargetType.Player;
```
Comment (QuiCM) [inline Terraria/Utilities/NPCUtils.cs:26]: Should probably make these float.MaxValue

File: Terraria/Utilities/NPCUtils.cs [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+
+			if (flags.HasFlag(NPCUtils.TargetSearchFlag.NPCs))
```
Comment (WhiteXZ) [inline Terraria/Utilities/NPCUtils.cs]: Should probably make these float.MaxValue

File: Terraria/Utilities/NPCUtils.cs [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+			{
```
Comment (WhiteXZ) [inline Terraria/Utilities/NPCUtils.cs]: Should probably make these float.MaxValue

File: Terraria/Utilities/SlotId.cs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+		}
+
+		public static bool operator !=(SlotId lhs, SlotId rhs)
+		{
+			return lhs.Value != rhs.Value;
+		}
+
+		internal SlotId ToActive(uint index)
+		{
+			uint num = 2147418112u & this.Key + 65536u;
```
Comment (QuiCM) [inline Terraria/Utilities/SlotId.cs]: KEY_MASK and KEY_INC should be used here
(`uint num = KEY_MASK & this.Key + KEY_INC`)

File: Terraria/Utilities/SlotId.cs [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+			return new SlotId(2147483648u | num | index);
```
Comment (QuiCM) [inline Terraria/Utilities/SlotId.cs]: ACTIVE_MASK
(`return new SlotId(ACTIVE_MASK | num | index)`)

File: Terraria/Utilities/SlotId.cs [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+		internal SlotId ToInactive(uint freeHead)
+		{
+			return new SlotId(this.Key | freeHead);
+		}
+
+		internal uint Index
+		{
+			get
+			{
+				return this.Value & 65535u;
```
Comment (QuiCM) [inline Terraria/Utilities/SlotId.cs]: INDEX_MASK
(`return this.Value & INDEX_MASK`)

File: Terraria/Utilities/SlotId.cs [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+			}
+		}
+
+		internal bool IsActive
+		{
+			get
+			{
+				return (this.Value & 2147483648u) != 0u && this.IsValid;
```
Comment (QuiCM) [inline Terraria/Utilities/SlotId.cs]: ACTIVE_MASK

File: Terraria/Utilities/SlotId.cs [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+			}
+		}
+
+		public bool IsValid
+		{
+			get
+			{
+				return (this.Value & 65535u) != 65535u;
```
Comment (QuiCM) [inline Terraria/Utilities/SlotId.cs]: INDEX_MASK

File: Terraria/Utilities/SlotId.cs [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+			}
+		}
+
+		internal uint Key
+		{
+			get
+			{
+				return this.Value & 2147418112u;
```
Comment (QuiCM) [inline Terraria/Utilities/SlotId.cs]: KEY_MASK

File: Terraria/Utilities/SlotId.cs [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+
+		private const uint KEY_INC = 65536u;
+
+		private const uint INDEX_MASK = 65535u;
+
+		private const uint ACTIVE_MASK = 2147483648u;
+
+		private const uint KEY_MASK = 2147418112u;
+
+		public static readonly SlotId Invalid = new SlotId(65535u);
```
Comment (QuiCM) [inline Terraria/Utilities/SlotId.cs]: INDEX_MASK

File: Terraria/Utilities/SlotVector.cs [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+			this._array = new SlotVector<T>.ItemPair[capacity];
+			this.Clear();
+		}
+		public SlotId Add(T value)
+		{
+			if (this._freeHead == 65535u)
+			{
+				return new SlotId(65535u);
+			}
+			uint freeHead = this._freeHead;
```
Comment (QuiCM) [inline Terraria/Utilities/SlotVector.cs:21]: MAX_INDEX should replace all 65535u in this class

File: Terraria/Utilities/SlotVector.cs [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+			this._count = 0;
+			this._freeHead = 0u;
+			uint num = 0u;
+			while ((ulong)num < (ulong)((long)(this._array.Length - 1)))
+			{
+				this._array[(int)((UIntPtr)num)] = new SlotVector<T>.ItemPair(default(T), new SlotId(num + 1u));
+				num += 1u;
+			}
+			this._array[this._array.Length - 1] = new SlotVector<T>.ItemPair(default(T), new SlotId(65535u));
+		}
```
Comment (QuiCM) [inline Terraria/Utilities/SlotVector.cs]: No idea why this is using UIntPtrs. Should probably check that out

File: Terraria/Utilities/SlotVector.cs [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+			return new SlotVector<T>.ItemPair(default(T), SlotId.Invalid);
+		}
+		public bool Has(SlotId id)
+		{
+			uint index = id.Index;
+			return index >= 0u && (ulong)index < (ulong)((long)this._array.Length) && this._array[(int)((UIntPtr)index)].Id.IsActive && !(id != this._array[(int)((UIntPtr)index)].Id);
+		}
+		public bool Has(int index)
+		{
+			return index >= 0 && index < this._array.Length && this._array[index].Id.IsActive;
```
Comment (QuiCM) [inline Terraria/Utilities/SlotVector.cs:64]: More UIntPtrs

File: Terraria/Utilities/SlotVector.cs [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+				{
+					throw new KeyNotFoundException();
+				}
+				return this._array[(int)((UIntPtr)index)].Value;
+			}
			set
+			{
+				uint index = id.Index;
+				if (index < 0u || (ulong)index >= (ulong)((long)this._array.Length))
+				{
```
Comment (QuiCM) [inline Terraria/Utilities/SlotVector.cs:153]: More UIntPtrs

File: Terraria/Utilities/SlotVector.cs [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+					throw new IndexOutOfRangeException();
+				}
+				if (!this._array[(int)((UIntPtr)index)].Id.IsActive || id != this._array[(int)((UIntPtr)index)].Id)
+				{
```
Comment (QuiCM) [inline Terraria/Utilities/SlotVector.cs:157]: And more...

File: Terraria/Utilities/SlotVector.cs [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+					throw new KeyNotFoundException();
+				}
+				this._array[(int)((UIntPtr)index)] = new SlotVector<T>.ItemPair(value, id);
+			}
+		}
+		private const uint MAX_INDEX = 65535u;
+		private SlotVector<T>.ItemPair[] _array;
+		private int _count;
+		private uint _freeHead;
```
Comment (QuiCM) [inline Terraria/Utilities/SlotVector.cs]: More...

File: Terraria/Utilities/SlotVector.cs [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+		private uint _usedSpaceLength;
+		public class Enumerator : IEnumerator<SlotVector<T>.ItemPair>, IDisposable, IEnumerator
+		{
			public Enumerator(SlotVector<T> slotVector)
+			{
```
Comment (QuiCM) [inline Terraria/Utilities/SlotVector.cs:177]: ...And more

File: Terraria/Utilities/SlotVector.cs [hunk 23]
```diff
@@ -0,0 +1,195 @@
...
+					if (this._slotVector.Has(this._index))
+					{
+						return true;
+					}
+				}
+				return false;
+			}
+			public void Reset()
+			{
+				this._index = -1;
```
Comment (QuiCM) [inline Terraria/Utilities/SlotVector.cs:195]: All these casts to long and ulong seem like they're unnecessary, maybe investigate what's going on there

**Review by QuiCM — CHANGES_REQUESTED** [review]
SlotVector and SlotID are weird.
There's also a lot of places where constants have been replaced with magic numbers (eg `float num = 3.40282347E+38f;` instead of `float num = float.MaxValue`)

**Review by QuiCM — APPROVED** [review]

**Pull request merged**